File: pkg/sdk/python/module.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from abc import ABC, abstractmethod
from typing import Dict, List, Any, Optional
import time
import json
import sys
import os
import signal
import traceback

# 导入自定义模块
from .logger import create_logger, Logger, LogLevel
from .config import ConfigHelper

class ModuleInfo:
    """模块信息类"""
    def __init__(self, id: str, name: str, version: str, description: str = "",
                 author: str = "", license: str = "", capabilities: List[str] = None,
                 supported_platforms: List[str] = None):
        self.id = id
        self.name = name
        self.version = version
        self.description = description
        self.author = author
        self.license = license
        self.capabilities = capabilities or []
        self.supported_platforms = supported_platforms or []
        self.language = "python"

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        return {
            "id": self.id,
            "name": self.name,
            "version": self.version,
            "description": self.description,
            "author": self.author,
            "license": self.license,
            "capabilities": self.capabilities,
            "supported_platforms": self.supported_platforms,
            "language": self.language
        }

class Request:
    """请求类"""
    def __init__(self, id: str, action: str, params: Dict[str, Any] = None,
                 metadata: Dict[str, str] = None, timeout: int = 30000):
        self.id = id
        self.action = action
        self.params = params or {}
        self.metadata = metadata or {}
        self.timeout = timeout

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Request':
        """从字典创建请求"""
        return cls(
            id=data.get("id", ""),
            action=data.get("action", ""),
            params=data.get("params", {}),
            metadata=data.get("metadata", {}),
            timeout=data.get("timeout", 30000)
        )

class Response:
    """响应类"""
    def __init__(self, id: str, success: bool = True, data: Dict[str, Any] = None,
                 error: Dict[str, Any] = None, metadata: Dict[str, str] = None):
        self.id = id
        self.success = success
        self.data = data or {}
        self.error = error
        self.metadata = metadata or {}

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        result = {
            "id": self.id,
            "success": self.success,
            "data": self.data,
            "metadata": self.metadata
        }
        if self.error:
            result["error"] = self.error
        return result

class Event:
    """事件类"""
    def __init__(self, id: str, type: str, source: str, data: Dict[str, Any] = None,
                 metadata: Dict[str, str] = None, timestamp: int = None):
        self.id = id
        self.type = type
        self.source = source
        self.data = data or {}
        self.metadata = metadata or {}
        self.timestamp = timestamp or int(time.time() * 1000)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Event':
        """从字典创建事件"""
        return cls(
            id=data.get("id", ""),
            type=data.get("type", ""),
            source=data.get("source", ""),
            data=data.get("data", {}),
            metadata=data.get("metadata", {}),
            timestamp=data.get("timestamp", int(time.time() * 1000))
        )

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        return {
            "id": self.id,
            "type": self.type,
            "source": self.source,
            "data": self.data,
            "metadata": self.metadata,
            "timestamp": self.timestamp
        }

class HealthStatus:
    """健康状态类"""
    def __init__(self, status: str = "healthy", details: Dict[str, Any] = None,
                 timestamp: int = None):
        self.status = status
        self.details = details or {}
        self.timestamp = timestamp or int(time.time() * 1000)

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        return {
            "status": self.status,
            "details": self.details,
            "timestamp": self.timestamp
        }

class Module(ABC):
    """模块基类"""

    def __init__(self):
        """初始化模块"""
        self.logger = create_logger(self.__class__.__name__, "info")
        self.config = {}
        self.config_helper = ConfigHelper({})
        self.start_time = time.time()

    @abstractmethod
    def init(self, config: Dict[str, Any]) -> None:
        """初始化模块"""
        self.config = config
        self.config_helper = ConfigHelper(config)

    @abstractmethod
    def start(self) -> None:
        """启动模块"""
        self.start_time = time.time()

    @abstractmethod
    def stop(self) -> None:
        """停止模块"""
        pass

    @abstractmethod
    def get_info(self) -> ModuleInfo:
        """获取模块信息"""
        pass

    @abstractmethod
    def handle_request(self, request: Request) -> Response:
        """处理请求"""
        pass

    @abstractmethod
    def handle_event(self, event: Event) -> bool:
        """处理事件"""
        pass

    def check_health(self) -> HealthStatus:
        """检查健康状态"""
        return HealthStatus(
            status="healthy",
            details={
                "uptime": time.time() - self.start_time
            },
            timestamp=int(time.time() * 1000)
        )

class BaseModule(Module):
    """基础模块实现"""

    def __init__(self, id: str, name: str, version: str, description: str = ""):
        """初始化基础模块"""
        super().__init__()
        self.id = id
        self.name = name
        self.version = version
        self.description = description
        self.author = ""
        self.license = ""
        self.capabilities = []
        self.supported_platforms = ["windows", "linux", "darwin"]

    def init(self, config: Dict[str, Any]) -> None:
        """初始化模块"""
        super().init(config)
        self.logger.info("初始化模块", id=self.id)

    def start(self) -> None:
        """启动模块"""
        super().start()
        self.logger.info("启动模块", id=self.id)

    def stop(self) -> None:
        """停止模块"""
        self.logger.info("停止模块", id=self.id)
        uptime = time.time() - self.start_time
        self.logger.info("运行时间", uptime=f"{uptime:.2f}秒")

    def get_info(self) -> ModuleInfo:
        """获取模块信息"""
        return ModuleInfo(
            id=self.id,
            name=self.name,
            version=self.version,
            description=self.description,
            author=self.author,
            license=self.license,
            capabilities=self.capabilities,
            supported_platforms=self.supported_platforms
        )

    def handle_request(self, request: Request) -> Response:
        """处理请求"""
        self.logger.info("处理请求", action=request.action)
        return Response(
            id=request.id,
            success=False,
            error={
                "code": "not_implemented",
                "message": f"未实现的操作: {request.action}"
            }
        )

    def handle_event(self, event: Event) -> bool:
        """处理事件"""
        self.logger.info("处理事件", type=event.type, source=event.source)
        return False

class ModuleRunner:
    """模块运行器"""

    def __init__(self, module: Module):
        """初始化运行器"""
        self.module = module
        self.logger = create_logger("ModuleRunner", "info")

    def run(self):
        """运行模块"""
        self.logger.info("模块运行器启动")

        # 设置信号处理
        signal.signal(signal.SIGINT, self._handle_signal)
        signal.signal(signal.SIGTERM, self._handle_signal)

        # 读取环境变量
        plugin_id = os.environ.get("KENNEL_PLUGIN_ID", "")
        config_json = os.environ.get("KENNEL_PLUGIN_CONFIG", "{}")

        try:
            # 解析配置
            config = json.loads(config_json)

            # 初始化模块
            self.logger.info("初始化模块")
            self.module.init(config)

            # 启动模块
            self.logger.info("启动模块")
            self.module.start()

            # 向标准输出写入就绪信息
            ready_info = {
                "status": "ready",
                "plugin_id": plugin_id,
                "info": self.module.get_info().to_dict()
            }
            print(f"KENNEL_PLUGIN_READY:{json.dumps(ready_info)}")
            sys.stdout.flush()

            # 进入命令处理循环
            self.command_loop()

        except Exception as e:
            self.logger.error(f"模块运行错误: {e}")
            self.logger.error(traceback.format_exc())
            error_info = {
                "status": "error",
                "plugin_id": plugin_id,
                "error": str(e)
            }
            print(f"KENNEL_PLUGIN_ERROR:{json.dumps(error_info)}")
            sys.stdout.flush()
            sys.exit(1)

    def _handle_signal(self, signum, frame):
        """处理信号"""
        self.logger.info(f"收到信号: {signum}")
        try:
            self.module.stop()
        except Exception as e:
            self.logger.error(f"停止模块错误: {e}")
        sys.exit(0)

    def command_loop(self):
        """命令处理循环"""
        self.logger.info("进入命令处理循环")

        try:
            while True:
                # 读取命令
                line = sys.stdin.readline().strip()
                if not line:
                    continue

                # 解析命令
                if line.startswith("KENNEL_COMMAND:"):
                    command_json = line[len("KENNEL_COMMAND:"):]
                    self.handle_command(command_json)
                elif line == "KENNEL_STOP":
                    self.logger.info("收到停止命令")
                    break
        except KeyboardInterrupt:
            self.logger.info("收到中断信号")
        except Exception as e:
            self.logger.error(f"命令处理错误: {e}")
            self.logger.error(traceback.format_exc())
        finally:
            # 停止模块
            self.logger.info("停止模块")
            try:
                self.module.stop()
            except Exception as e:
                self.logger.error(f"停止模块错误: {e}")
                self.logger.error(traceback.format_exc())

    def handle_command(self, command_json: str):
        """处理命令"""
        try:
            command = json.loads(command_json)
            command_type = command.get("type", "")
            command_data = command.get("data", {})

            if command_type == "request":
                # 处理请求
                request = Request.from_dict(command_data)
                self.logger.info(f"处理请求: {request.action}")
                response = self.module.handle_request(request)
                self.send_response(response)
            elif command_type == "event":
                # 处理事件
                event = Event.from_dict(command_data)
                self.logger.info(f"处理事件: {event.type}")
                success = self.module.handle_event(event)
                self.send_event_response(event.id, success)
            elif command_type == "health_check":
                # 健康检查
                self.logger.info("健康检查")
                health = self.module.check_health()
                self.send_health_response(health)
            else:
                self.logger.warning(f"未知命令类型: {command_type}")
        except Exception as e:
            self.logger.error(f"处理命令错误: {e}")
            self.logger.error(traceback.format_exc())
            self.send_error_response(str(e))

    def send_response(self, response: Response):
        """发送响应"""
        response_json = json.dumps(response.to_dict())
        print(f"KENNEL_RESPONSE:{response_json}")
        sys.stdout.flush()

    def send_event_response(self, event_id: str, success: bool):
        """发送事件响应"""
        response = {
            "event_id": event_id,
            "success": success
        }
        response_json = json.dumps(response)
        print(f"KENNEL_EVENT_RESPONSE:{response_json}")
        sys.stdout.flush()

    def send_health_response(self, health: HealthStatus):
        """发送健康响应"""
        response = health.to_dict()
        response_json = json.dumps(response)
        print(f"KENNEL_HEALTH_RESPONSE:{response_json}")
        sys.stdout.flush()

    def send_error_response(self, error: str):
        """发送错误响应"""
        response = {
            "error": error
        }
        response_json = json.dumps(response)
        print(f"KENNEL_ERROR:{response_json}")
        sys.stdout.flush()

def run_module(module: Module):
    """运行模块"""
    runner = ModuleRunner(module)
    runner.run()

File: pkg/sdk/go/module.go
package sdk

import (
	"context"
	"fmt"
	"os"
	"time"

	goplugin "github.com/hashicorp/go-plugin"
	"github.com/lomehong/kennel/pkg/core/plugin"
	"github.com/lomehong/kennel/pkg/logging"
	pluginLib "github.com/lomehong/kennel/pkg/plugin"
)

// BaseModule 基础模块实现
type BaseModule struct {
	ID          string
	Name        string
	Version     string
	Description string
	Author      string
	License     string
	Logger      logging.Logger
	Config      map[string]interface{}
	StartTime   time.Time
}

// NewBaseModule 创建基础模块
func NewBaseModule(id, name, version, description string) *BaseModule {
	// 创建默认日志记录器
	logConfig := logging.DefaultLogConfig()
	logConfig.Level = logging.LogLevelInfo
	logConfig.Output = logging.LogOutputStdout
	logger, err := logging.NewEnhancedLogger(logConfig)
	if err != nil {
		// 如果创建失败，创建一个基本的日志记录器
		logConfig.Output = logging.LogOutputStdout
		logConfig.Format = logging.LogFormatText
		logger, _ = logging.NewEnhancedLogger(logConfig)
	}

	return &BaseModule{
		ID:          id,
		Name:        name,
		Version:     version,
		Description: description,
		Logger:      logger.Named(id),
		Config:      make(map[string]interface{}),
		StartTime:   time.Now(),
	}
}

// Init 初始化模块
func (m *BaseModule) Init(ctx context.Context, config *plugin.ModuleConfig) error {
	m.Logger.Info("初始化模块", "id", m.ID)
	m.Config = config.Settings
	return nil
}

// Start 启动模块
func (m *BaseModule) Start() error {
	m.Logger.Info("启动模块", "id", m.ID)
	m.StartTime = time.Now()
	return nil
}

// Stop 停止模块
func (m *BaseModule) Stop() error {
	m.Logger.Info("停止模块", "id", m.ID)
	uptime := time.Since(m.StartTime)
	m.Logger.Info("运行时间", "uptime", uptime.String())
	return nil
}

// GetInfo 获取模块信息
func (m *BaseModule) GetInfo() plugin.ModuleInfo {
	return plugin.ModuleInfo{
		ID:                 m.ID,
		Name:               m.Name,
		Version:            m.Version,
		Description:        m.Description,
		Author:             m.Author,
		License:            m.License,
		Capabilities:       []string{},
		SupportedPlatforms: []string{"windows", "linux", "darwin"},
		Language:           "go",
	}
}

// HandleRequest 处理请求
func (m *BaseModule) HandleRequest(ctx context.Context, req *plugin.Request) (*plugin.Response, error) {
	m.Logger.Info("处理请求", "action", req.Action)
	return &plugin.Response{
		ID:      req.ID,
		Success: false,
		Error: &plugin.ErrorInfo{
			Code:    "not_implemented",
			Message: fmt.Sprintf("未实现的操作: %s", req.Action),
		},
	}, nil
}

// HandleEvent 处理事件
func (m *BaseModule) HandleEvent(ctx context.Context, event *plugin.Event) error {
	m.Logger.Info("处理事件", "type", event.Type, "source", event.Source)
	return nil
}

// CheckHealth 检查健康状态
func (m *BaseModule) CheckHealth() plugin.HealthStatus {
	return plugin.HealthStatus{
		Status: "healthy",
		Details: map[string]interface{}{
			"uptime": time.Since(m.StartTime).String(),
		},
		Timestamp: time.Now().UnixNano() / int64(time.Millisecond),
	}
}

// RunModule 运行模块
func RunModule(module plugin.Module) {
	// 设置环境变量，确保插件使用正确的 Magic Cookie
	os.Setenv("APPFRAMEWORK_PLUGIN", "appframework")

	// 启动模块
	if err := module.Start(); err != nil {
		fmt.Fprintf(os.Stderr, "启动模块失败: %v\n", err)
		os.Exit(1)
	}

	// 创建适配器
	adapter := &ModuleAdapter{
		Module: module,
	}

	// 启动插件服务
	goplugin.Serve(&goplugin.ServeConfig{
		HandshakeConfig: goplugin.HandshakeConfig{
			ProtocolVersion:  1,
			MagicCookieKey:   "APPFRAMEWORK_PLUGIN",
			MagicCookieValue: "appframework",
		},
		Plugins: map[string]goplugin.Plugin{
			"module": &pluginLib.ModulePlugin{Impl: adapter},
		},
		GRPCServer: goplugin.DefaultGRPCServer,
	})
}

// GetConfigString 获取配置字符串
func GetConfigString(config map[string]interface{}, key, defaultValue string) string {
	if value, ok := config[key]; ok {
		if strValue, ok := value.(string); ok {
			return strValue
		}
	}
	return defaultValue
}

// GetConfigInt 获取配置整数
func GetConfigInt(config map[string]interface{}, key string, defaultValue int) int {
	if value, ok := config[key]; ok {
		switch v := value.(type) {
		case int:
			return v
		case int64:
			return int(v)
		case float64:
			return int(v)
		}
	}
	return defaultValue
}

// GetConfigBool 获取配置布尔值
func GetConfigBool(config map[string]interface{}, key string, defaultValue bool) bool {
	if value, ok := config[key]; ok {
		if boolValue, ok := value.(bool); ok {
			return boolValue
		}
	}
	return defaultValue
}

// GetConfigMap 获取配置映射
func GetConfigMap(config map[string]interface{}, key string) map[string]interface{} {
	if value, ok := config[key]; ok {
		if mapValue, ok := value.(map[string]interface{}); ok {
			return mapValue
		}
	}
	return make(map[string]interface{})
}

// GetConfigSlice 获取配置切片
func GetConfigSlice(config map[string]interface{}, key string) []interface{} {
	if value, ok := config[key]; ok {
		if sliceValue, ok := value.([]interface{}); ok {
			return sliceValue
		}
	}
	return make([]interface{}, 0)
}

// GetConfigStringSlice 获取配置字符串切片
func GetConfigStringSlice(config map[string]interface{}, key string) []string {
	slice := GetConfigSlice(config, key)
	result := make([]string, 0, len(slice))
	for _, item := range slice {
		if strItem, ok := item.(string); ok {
			result = append(result, strItem)
		}
	}
	return result
}
